### ISSUE TITLE
Box raises an AttributeError on nonexistent key, should raise a KeyError (or custom error)

### ISSUE THREAD

**Issue by justiniso** [cdgriffith/Box#37]
When accessing a key which does not exist, an AttributeError will be raised. This sends the wrong message (in my opinion), implying that the user is attempting to access an attribute on the `Box` object which does not exist. When the true behavior is just syntactic sugar over a dict object, a KeyError seems more appropriate. I get that this is a philosophical opinion, so I could certainly be convinced that AttributeError is appropriate. 

Repro in python repl:

```python
Python 2.7.10 (default, Aug 22 2015, 20:33:39)
[GCC 4.2.1 Compatible Apple LLVM 7.0.0 (clang-700.0.59.1)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from box import Box
>>>
>>> b = Box({})
>>> b.attr
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/jiso/projects/jwplayer-python/venv/lib/python2.7/site-packages/box.py", line 385, in __getattr__
    raise err
AttributeError: 'Box' object has no attribute 'attr'
```

Actual: in above example, AttributeError is raised
Expected: KeyError is raised

**Comment by cdgriffith** [comment]
Hey @justiniso, thanks for the well laid out suggestion. this actually has been an issue I flipped back and forth on, and haven't really addressed why it is what it currently is anywhere (and also is still open to change for later versions). 

I am going to spill out my thoughts on the matter here, what I think the best solution going forward could be, and leave this issue open for a bit for discussion from others as well.

Currently Box v3 will raise KeyError when being accessed as a traditional dictionary and a key does not exist, and an AttributeError if accessed in dot notation. This is following with the idea that you know which method you are using when you write it, so you know which error to except (as it seems odd to `except KeyError` for an attribute). When I decided on this route I was trying to only use built-in errors to be as drop in replaceable to a dictionary. 

What I forgot to take into account with v3 is that there is now a custom BoxError, which means anyone using the library should expect that any errors raised by this library will be caught with that exception. 

My new train of thought is that a custom access error that is a subclass of all three should be used. 

```python
class BoxKeyError(BoxError, KeyError, AttributeError): 
    """Key does not exist"""
``` 

That way catching any of those would be valid, and would be backwards compatible. 

Thoughts?
